File: web/diagnostic.js
/* Corporate Distress Assessment screen.
 *
 * A diagnosis of filings already made. Three rules hold throughout, and each
 * is structural rather than a line in the disclaimer:
 *
 *   Nothing is called "current". Every state carries the period it came from
 *   and, where the filing is old, how old. Diebold at a 2024 prediction date
 *   is reading a 2022 balance sheet, and a reader must see that.
 *
 *   Reported and calculated are separate tables. A filed figure can be checked
 *   against the document; a ratio can only be checked by trusting the formula.
 *   Mixing them would present both as equally verifiable.
 *
 *   Zone names describe a condition, never an outcome. "Elevated stress" is a
 *   statement about a balance sheet. "Likely to fail" is a forecast, and this
 *   screen has no basis for one.
 *
 * Colour follows the reserved status scale: amber and crimson appear only
 * against a limit the filed figures actually breach, never as decoration.
 */

const ZONE_COLOR = {
  stable:       '#0ca30c',
  monitored:    '#fab219',
  elevated:     '#ec835a',
  severe:       '#d03b3b',
  insufficient: '#8794a8',
};
const ZONE_ORDER = ['stable', 'monitored', 'elevated', 'severe'];

const money = v => {
  if (v === null || v === undefined) return '—';
  const a = Math.abs(v);
  if (a >= 1e9) return (v / 1e9).toFixed(2) + 'bn';
  if (a >= 1e6) return (v / 1e6).toFixed(1) + 'm';
  if (a >= 1e3) return (v / 1e3).toFixed(0) + 'k';
  return v.toFixed(0);
};
const ratio = v => (v === null || v === undefined) ? '—'
  : (Math.abs(v) >= 1000 ? v.toExponential(2) : v.toFixed(2));

/* Segmented linear gauge. The active segment is filled; the rest are outlines,
 * so the reading is legible without relying on hue alone. */
function stressGauge(d) {
  const active = ZONE_ORDER.indexOf(d.zone);
  const segs = ZONE_ORDER.map((z, i) => {
    const on = i === active;
    const col = ZONE_COLOR[z];
    const lbl = d.zones.find(x => x.key === z)?.label || z;
    return `<div class="seg${on ? ' on' : ''}" style="--c:${col}">
        <div class="seg-bar"></div><div class="seg-lbl">${lbl}</div></div>`;
  }).join('');
  const note = d.zones.find(x => x.key === d.zone)?.note || '';
  const breached = d.calculated.filter(c => c.breached).length;
  const computable = d.calculated.filter(c => c.computable).length;
  return `<div class="card">
    <h3>Condition as of ${d.period_end || d.as_of}</h3>
    <div class="gauge">${segs}</div>
    <div class="gauge-read">
      <span class="zone-chip" style="--c:${ZONE_COLOR[d.zone]}">${d.zone_label}</span>
      <span class="gauge-sub">${breached} of ${computable} conventional limits breached
        in the figures filed for the period ending ${d.period_end || '—'}.</span>
    </div>
    <p class="note">${note} This is a reading of filings already made, assessed
      as of ${d.as_of}. It is not a projection.</p>
    ${d.stale ? `<p class="note warnnote">The newest usable filing is
      ${d.filing_age_days} days old. Every figure on this screen describes that
      vintage, and the position may since have changed in either direction.</p>` : ''}
  </div>`;
}

/* Reported facts and calculated indices, deliberately side by side but never
 * merged -- the distinction is the point. */
function factorMatrix(d) {
  const rep = d.reported.map(r => `<tr><td>${r.label}</td>
      <td class="num">${r.computable ? money(r.value) : '<span class="dim">not tagged</span>'}</td></tr>`).join('');
  const calc = d.calculated.map(c => `<tr>
      <td>${c.label}</td>
      <td class="num ${c.breached ? 'sev' : ''}">${c.computable ? ratio(c.value) : '<span class="dim">n/a</span>'}</td>
      <td class="num dim">${c.threshold}</td>
      <td>${c.breached ? '<span class="pill breach">breached</span>'
          : c.computable ? '<span class="pill within">within</span>' : ''}</td></tr>`).join('');
  return `<div class="grid2">
    <div class="card">
      <h3>Reported figures <span class="tier backtested">as filed</span></h3>
      <table><tr><th>Line item</th><th style="text-align:right">Value (USD)</th></tr>${rep}</table>
      <p class="note">Taken directly from the company's own filing for the period
        ending ${d.period_end || '—'}. Each can be checked against the document.
        &ldquo;Not tagged&rdquo; means the filer did not report that item in
        machine-readable form, which is itself a finding.</p>
    </div>
    <div class="card">
      <h3>Solvency factors <span class="tier context-only">calculated</span></h3>
      <table><tr><th>Measure</th><th style="text-align:right">Value</th>
        <th style="text-align:right">Limit</th><th>State</th></tr>${calc}</table>
      <p class="note">Arithmetic performed on the figures at left, against
        conventional credit levels rather than thresholds fitted to any outcome.
        Fitting them would make this a model of past bankruptcies rather than a
        reading of the balance sheet.</p>
    </div>
  </div>`;
}

/* Vertical timeline: the diagnosis as it would have read at each earlier
 * period, recomputed from what was public then. */
function filingTimeline(d) {
  if (!d.timeline.length) return '';
  const rows = d.timeline.map(t => `
    <div class="tl-row">
      <div class="tl-date">${t.as_of}</div>
      <div class="tl-mark"><span style="--c:${ZONE_COLOR[t.zone]}"></span></div>
      <div class="tl-body">
        <div class="tl-zone" style="color:${ZONE_COLOR[t.zone]}">
          ${(d.zones.find(z => z.key === t.zone) || {}).label || t.zone}</div>
        <div class="tl-note">${t.breached} of ${t.computable} limits breached
          in the figures public by that date</div>
      </div>
    </div>`).join('');
  return `<div class="card">
    <h3>Retrospective filing timeline</h3>
    <div class="timeline">${rows}
      <div class="tl-row current">
        <div class="tl-date">${d.period_end || d.as_of}</div>
        <div class="tl-mark"><span style="--c:${ZONE_COLOR[d.zone]}"></span></div>
        <div class="tl-body">
          <div class="tl-zone" style="color:${ZONE_COLOR[d.zone]}">${d.zone_label}</div>
          <div class="tl-note">latest period visible as of ${d.as_of}</div>
        </div>
      </div>
    </div>
    <p class="note">Each reading was recomputed from filings public at that date,
      not by applying today's view to older numbers. That is what makes the
      series a record rather than a reconstruction.</p>
  </div>`;
}

function companyHeader(d) {
  return `<div class="card idhead">
    <div>
      <div class="idname">${d.name || 'CIK ' + d.cik}</div>
      <div class="idmeta">
        <span>CIK <b>${String(d.cik).padStart(10, '0')}</b></span>
        ${d.ticker ? `<span>Ticker <b>${d.ticker}</b></span>` : ''}
        ${d.exchange ? `<span>Listed <b>${d.exchange}</b></span>` : ''}
        ${d.state ? `<span>Incorporated <b>${d.state}</b></span>` : ''}
      </div>
      <div class="idsic">SIC ${d.sic} &mdash; ${d.sic_description}</div>
    </div>
    <div class="idzone">
      <div class="idzone-l">Assessed as of</div>
      <div class="idzone-v">${d.as_of}</div>
    </div>
  </div>`;
}

function renderDiagnostic(host, d) {
  if (d.error) {
    const list = (d.candidates || []).map(c => `<tr class="clickable" data-pick="${c.cik}">
        <td>${c.name}</td><td>${c.ticker}</td><td class="num">${c.cik}</td></tr>`).join('');
    host.innerHTML = `<div class="card blocked"><h3>No diagnosis</h3>
      <p class="err">${d.error}</p>${list ? `<table style="margin-top:12px">
      <tr><th>Company</th><th>Ticker</th><th style="text-align:right">CIK</th></tr>
      ${list}</table>` : ''}</div>`;
    return;
  }
  host.innerHTML = companyHeader(d) + stressGauge(d) + factorMatrix(d) + filingTimeline(d)
    + `<div class="card disclaimer">
        This tool provides an analytical diagnosis of past and present financial
        filings. It does not predict future performance, market viability, or
        guarantee bankruptcy outcomes.
      </div>`;
}

File: agents/diagnostic.py
"""A corporate distress diagnostic: past and present states only.

The screen this serves is a *diagnosis of filings already made*, not a view on
what happens next, and three choices in here enforce that rather than merely
asserting it in a disclaimer.

**Every state is dated.** There is no "current" reading. A filer's latest
visible annual period may be eighteen months before the date being asked
about, and labelling that "current" implies live data nobody has. The vintage
travels with the number.

**Reported facts and calculated indices are separated.** ``total_assets`` is a
figure the company filed. ``altman_z_double_prime`` is arithmetic performed on
several such figures. A reader can check the first against the filing and can
only check the second by trusting the formula, so the two are kept in
different columns rather than mixed into one table of "metrics".

**The historical series is recomputed, not remembered.** Each point on the
timeline is the diagnostic as it would have read on that date, using only
filings public by then. It is not today's model applied to old numbers, which
would import knowledge the period did not have.

No zone here is named for an outcome. "Elevated stress" describes a balance
sheet; "likely to fail" describes a future, and this module has no basis for
the second.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Any

#: Diagnostic zones, ordered. Named for the condition observed, never for a
#: consequence -- "monitored" is a state a reader can verify, "at risk of
#: bankruptcy" is a claim about the future.
ZONES = (
    ("stable", "Stable", "No threshold breached in the filed figures."),
    ("monitored", "Monitored", "One or more measures near a conventional limit."),
    ("elevated", "Elevated stress", "Several measures past conventional limits."),
    ("severe", "Severe stress", "Core solvency measures breached together."),
    ("insufficient", "Insufficient data", "Too few figures to form a diagnosis."),
)

#: Our five-level signal maps onto the diagnostic zones one-for-one.
SIGNAL_TO_ZONE = {
    "healthy": "stable",
    "watch": "monitored",
    "elevated_risk": "elevated",
    "severe_risk": "severe",
    "insufficient_evidence": "insufficient",
}

#: Figures the company filed. A reader can open the 10-K and check these.
REPORTED = (
    ("total_assets", "Total assets"),
    ("total_liabilities", "Total liabilities"),
    ("current_assets", "Current assets"),
    ("current_liabilities", "Current liabilities"),
    ("revenue", "Revenue"),
    ("net_income", "Net income"),
    ("cash_from_operations", "Cash from operations"),
)

#: Arithmetic performed on the above. Checkable only by trusting the formula,
#: so shown separately.
CALCULATED = (
    ("current_ratio", "Current ratio", "1.00"),
    ("quick_ratio", "Quick ratio", "1.00"),
    ("liabilities_to_assets", "Liabilities to assets", "0.70"),
    ("debt_to_assets", "Debt to assets", "0.60"),
    ("interest_coverage", "Interest coverage", "1.50"),
    ("ocf_to_debt", "Operating cash flow to debt", "0.10"),
    ("net_margin", "Net margin", "0.00"),
    ("altman_z_double_prime", "Altman Z'' score", "1.10"),
)


@dataclass
class Reading:
    """One figure, with whether it was filed or derived."""

    key: str
    label: str
    value: float | None
    computable: bool
    threshold: str = ""
    breached: bool = False


@dataclass
class TimelinePoint:
    """The diagnosis as it would have read on one past date."""

    as_of: date
    period_end: date | None
    zone: str
    breached: int
    computable: int


@dataclass
class Diagnostic:
    cik: int
    name: str = ""
    ticker: str = ""
    exchange: str = ""
    sic: str = ""
    sic_description: str = ""
    state: str = ""
    as_of: date | None = None
    period_end: date | None = None
    filing_age_days: int | None = None
    zone: str = "insufficient"
    reported: list[Reading] = field(default_factory=list)
    calculated: list[Reading] = field(default_factory=list)
    timeline: list[TimelinePoint] = field(default_factory=list)

    @property
    def zone_label(self) -> str:
        return next((lbl for k, lbl, _ in ZONES if k == self.zone), self.zone)

    @property
    def stale(self) -> bool:
        """Whether the newest usable filing is old enough to say so."""
        return self.filing_age_days is not None and self.filing_age_days > 456


def _breaches(key: str, value: float) -> bool:
    """Whether a calculated figure sits past its conventional limit.

    Conventional levels from credit practice, not fitted to our outcomes --
    fitting them would make this a model of our labels rather than a reading
    of the balance sheet.
    """
    limits = {
        "current_ratio": lambda v: v < 1.0,
        "quick_ratio": lambda v: v < 1.0,
        "liabilities_to_assets": lambda v: v > 0.7,
        "debt_to_assets": lambda v: v > 0.6,
        "interest_coverage": lambda v: v < 1.5,
        "ocf_to_debt": lambda v: v < 0.1,
        "net_margin": lambda v: v < 0.0,
        "altman_z_double_prime": lambda v: v < 1.1,
    }
    test = limits.get(key)
    return bool(test and test(value))


def build(
    cik: int,
    as_of: date,
    facts: Any,
    submissions: dict[str, Any] | None = None,
    history: int = 6,
) -> Diagnostic:
    """Assemble the diagnostic for one filer at one date."""
    from compute import lineitems
    from compute.lineitems import FactIndex, annual_period_ends
    from compute.ratios import compute_metric
    from data.facts import as_of_view

    sub = submissions or {}
    out = Diagnostic(
        cik=cik,
        name=str(sub.get("name") or ""),
        ticker=(sub.get("tickers") or [""])[0] if sub.get("tickers") else "",
        exchange=(sub.get("exchanges") or [""])[0] if sub.get("exchanges") else "",
        sic=str(sub.get("sic") or ""),
        sic_description=str(sub.get("sicDescription") or ""),
        state=str(sub.get("stateOfIncorporation") or ""),
        as_of=as_of,
    )

    view = FactIndex(as_of_view(facts, as_of))
    ends = annual_period_ends(view)
    if not ends:
        return out
    latest = ends[0]
    out.period_end = latest
    out.filing_age_days = (as_of - latest).days

    for key, label in REPORTED:
        # resolve() returns the fact reference the filer actually filed --
        # a value a reader can find in the document, unlike a computed ratio.
        ref = lineitems.resolve(key, view, latest)
        out.reported.append(
            Reading(key, label, float(ref.value) if ref else None, ref is not None)
        )

    breached = 0
    for key, label, limit in CALCULATED:
        cv = compute_metric(key, view, latest)
        ok = cv.is_defined
        value = float(cv.value) if ok else None
        hit = bool(ok and _breaches(key, value))
        breached += hit
        out.calculated.append(Reading(key, label, value, ok, limit, hit))

    computable = sum(1 for r in out.calculated if r.computable)
    out.zone = _zone_from(breached, computable)

    # Recompute the diagnosis at each earlier period end, using only what was
    # public by then. Not today's reading applied backwards.
    for end in ends[1 : history + 1]:
        past_view = FactIndex(as_of_view(facts, end))
        past_ends = annual_period_ends(past_view)
        if not past_ends:
            continue
        pb = pc = 0
        for key, _lbl, _lim in CALCULATED:
            cv = compute_metric(key, past_view, past_ends[0])
            if cv.is_defined:
                pc += 1
                pb += _breaches(key, float(cv.value))
        out.timeline.append(
            TimelinePoint(end, past_ends[0], _zone_from(pb, pc), pb, pc)
        )
    out.timeline.reverse()
    return out


def _zone_from(breached: int, computable: int) -> str:
    """Zone from how many conventional limits the filed figures breach.

    Counting breaches rather than fitting a score keeps this a reading of the
    balance sheet. A fitted threshold would encode our own outcome labels and
    make the gauge a quiet prediction.
    """
    if computable < 3:
        return "insufficient"
    share = breached / computable
    if breached == 0:
        return "stable"
    if share < 0.25:
        return "monitored"
    if share < 0.5:
        return "elevated"
    return "severe"
